src/serdes/binary-parser.ts

```diff
@@ -107,10 +107,26 @@ class BinaryParser {
    * @return Field ordinal
    */
   readFieldOrdinal(): number {
-    const tagByte = this.readUInt8();
-    const type = (tagByte & 0xf0) >>> 4 || this.readUInt8();
-    const nth = tagByte & 0x0f || this.readUInt8();
-    return (type << 16) | nth;
+    let type = this.readUInt8();
+    let field = type & 15;
+    type >>= 4;
+
+
+    if(type === 0) {
+      type = this.readUInt8();
+      if(type === 0 || type < 16) {
+        throw new Error("Cannot read FieldOrdinal, type_code out of range");
+      }
+    }
+    
+    if(field === 0) {
+      field = this.readUInt8();
+      if(field === 0 || field < 16) {
+        throw new Error("Cannot read FieldOrdinal, field_code out of range")
+      }
+    }
+
+    return (type << 16) | field;
```

Comment:
Same logic as in (rippled decode FieldID)[https://github.com/seelabs/rippled/blob/cecc0ad75849a1d50cc573188ad301ca65519a5b/src/ripple/protocol/impl/Serializer.cpp#L484-L509]